### PR TITLE
Tn clean upsample

### PR DESCRIPTION
**Pull request by yzhang123**
* added cleaning script
* added upsampling scrpit
* modified input processing , e.g. to also add space between alpha and numerical characters
* added electronic wfst as part of preprocessing to inference script

### CODE REVIEW

File: examples/nlp/duplex_text_normalization/data_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+        o_tokens = wordninja.split(o)
+        o = ' '.join(o_tokens)
+
+    return o
+
+
+def int2digits(digits: str):
+    """
+    Verbalizes integer digit by digit, e.g. "12,000.12" -> "one two zero zero zero point one two"
+    It can also take in a string that has an integer as prefix and outputs only the verbalized part of that, e.g. "12 kg" -> "one two"
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/data_preprocessing.py:171]: its only the prefix number as discussed. Comment in 2nd line

File: examples/nlp/duplex_text_normalization/data_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This script can be used to clean the splits of English Google Text Normalization dataset
+for better training performance.
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/data_preprocessing.py:17]: done

File: examples/nlp/duplex_text_normalization/data_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        written = re.sub("([0-9])\s?''", '\\1"', written)
+
+    spoken = process_url(spoken)
+
+    if cls in ["TELEPHONE", "DIGIT", "MEASURE", "DECIMAL", "MONEY", "ADDRESS"]:
+        spoken = re.sub(" o ", " zero ", spoken)
+        spoken = re.sub(" o ", " zero ", spoken)
+        spoken = re.sub(" o ", " zero ", spoken)
+        spoken = re.sub(" o ", " zero ", spoken)
+        spoken = re.sub(" o ", " zero ", spoken)
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/data_preprocessing.py:286]: to make sure overlapping matches are also substituted, 2x is enough, changed

File: examples/nlp/duplex_text_normalization/data_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    "--max_denominator_length",
+    default=3,
+    type=int,
+    help='Maximum number of digits for denominators that are allowed. Beyond this, the denominator is verbalized digit by digit.',
+)
+args = parser.parse_args()
+
+engine = inflect.engine()
+
+number_verbalizations = list(range(0, 20)) + list(range(20, 100, 10))
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/data_preprocessing.py:74]: done

File: examples/nlp/duplex_text_normalization/data_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        s_words = spoken.split()
+        for i, x in enumerate(s_words):
+            if x in number_verbalizations:
+                break
+        spoken = " ".join(s_words[:i]) + " " + number
+    elif cls == "DECIMAL":
+        res = []
+        for i, x in enumerate(written):
+            if i == 0 and x == "-":
+                res.append("minus")
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/data_preprocessing.py:342]: this is only for english, will move file to english folder

File: examples/nlp/duplex_text_normalization/data_split.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        fp = join(args.data_dir, fn)
+        if not isfile(fp):
+            continue
+        if not fn.startswith('output'):
+            continue
+
+        # Determine the current split
+        split_nb = int(fn.split('-')[1])
+        if split_nb < 90:
+            cur_split = "train"
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/data_split.py:113]: will leave this as discussed

File: examples/nlp/duplex_text_normalization/upsample.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+from collections import defaultdict
+from typing import List
+
+import numpy as np
+import regex as re
+
+parser = ArgumentParser(description="English Text Normalization upsampling")
+parser.add_argument("--input_dir", required=True, type=str, help='Path to input directory with preprocessed data')
+parser.add_argument("--output_file", required=True, type=str, help='Path to output file')
+parser.add_argument("--thresh", default=2000, type=int, help='minimum number per pattern')
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/duplex_text_normalization/upsample.py:47]: ok

**Review by ekmb — APPROVED** [review]

**Pull request merged**